iwyu.cc

```diff
@@ -2043,21 +2043,87 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
+    const Type* rhs_type = nullptr;
+    if (expr->getNumArgs() == 2)
+      rhs_type = expr->getArg(1)->getType().getTypePtr();
+
+    // Most of the binary type traits require the types to be complete types or
+    // arrays of unknown bound for compilable code. It means that compilers
+    // accept pointers or references to fwd-declared types. However, in certain
+    // cases it may cause a UB (C++20 [meta.rqmts]p5) because the full
+    // pointed-to type is needed to determine the trait value correctly, so it's
+    // better for IWYU to require the full type info in such situations. It is
+    // important not to suggest a full type when it is redundant for computing
+    // the correct value.
+    switch (expr->getTrait()) {
+      case TypeTrait::BTT_IsSame:
+        return true;
+      case TypeTrait::BTT_IsAssignable:
+      case TypeTrait::BTT_IsNothrowAssignable:
+        // TODO(bolshakov): require complete argument types here when
+        // CanForwardDeclareType stops doing it.
+        if (RemoveReference(lhs_type)->isRecordType()) {
+          // If the first type is a class/struct/union type, it may have
+          // a user-defined operator= which accepts the second type or its base
+          // class by reference or by pointer.
+          ReportTypeUse(CurrentLoc(), lhs_type, DerefKind::RemoveRefs);
+          ReportTypeUse(CurrentLoc(), rhs_type, DerefKind::RemoveRefsAndPtr);
+          // TODO(bolshakov): report a 3rd type involved in the conversion, if
+          // present. See an example in
+          // github.com/include-what-you-use/include-what-you-use/pull/745.
+        } else if (IsReferenceToModifiableLValue(lhs_type)) {
+          // Scalar types can be assigned to only if they're modifiable lvalues.
+          const Type* lhs_deref_type = RemoveReference(lhs_type);
+          const Type* rhs_deref_type = RemoveReference(rhs_type);
+          // For derived-to-base pointer conversion to be allowed, the derived
+          // type should be complete.
+          // TODO(bolshakov): what about member pointers?
+          if (DerivedToBasePtrConvertible(rhs_deref_type, lhs_deref_type)) {
+            ReportTypeUse(CurrentLoc(), rhs_type, DerefKind::RemoveRefsAndPtr);
+            return true;
+          }
+          // Otherwise, report rhs record types (ReportTypeUse doesn't report
+          // other types) because they may have a conversion operator to the lhs
+          // type.
+          ReportTypeUse(CurrentLoc(), rhs_type, DerefKind::RemoveRefs);
+        }
+        return true;
+      case TypeTrait::BTT_IsTriviallyAssignable: {
+        // Trivial assignment doesn't involve any user-defined operators but may
+        // involve a derived-to-base conversion.
+        const Type* lhs_deref_type = RemoveReference(lhs_type);
+        if (lhs_deref_type->isStructureOrClassType()) {
+          // Rhs may be the same or derived class. In both cases, its type info
```

Comment:
All-upper `RHS`, please.

Reply:
Hmm, or maybe that makes the other all-lower `rhs` look weird. Maybe compromise with `The rhs...`?

Reply:
The lowercase abbreviations seem to be more common in variable names ([an example in the C++ Core Guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c83-for-value-like-types-consider-providing-a-noexcept-swap-function)), but it may because the uppercase ones would look like macro names...

Reply:
Yeah, let's go with `The rhs may be the same...`, to avoid the casing challenge.